### PR TITLE
add basic rate limiting plugin

### PR DESCRIPTION
**Pull request by mrwanny**
### Summary

add basic rate limiting plugin to kong

### Checklist

- [x] The Pull Request has tests
- [ ] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* add basic rate limiting plugin to kong

[FTI-5118](https://konghq.atlassian.net/browse/FTI-5118)

[FTI-5118]: https://konghq.atlassian.net/browse/FTI-5118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/plugins/basic-rate-limiting/handler.lua [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+local ngx = ngx
+local shm = ngx.shared.kong_rate_limiting_counters
+
+local BasicRateLimitingHandler = {}
+
+BasicRateLimitingHandler.VERSION = kong_meta.version
+BasicRateLimitingHandler.PRIORITY = 911
+
+function BasicRateLimitingHandler:access(conf)
+    local counter = shm:incr("BasicRateLimiting",1,0,60)
```
Comment (catbro666) [inline kong/plugins/basic-rate-limiting/handler.lua:14]: In the original rate limiting plugin, the key is based on the current timestamp, so every window aligns with the physical time. In the current implementation, they share the same key, so a window starts when the first request comes in. Not a big difference, I think the current way maybe even better.

- It makes the code simpler. Only one key is needed for every window size.
- The limit is more strict. For the original implementation, the rate could be doubled at the border of two windows.
- It makes the tests less flaky as they don't rely on the start time of test any more.

File: spec/03-plugins/38-basic-rate-limiting/02-access_spec.lua [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    end)
+
+    after_each(function()
+      if proxy_client then
+        proxy_client:close()
+      end
+    end)
+
+    describe("basic-rate-limiting", function()
+      it("only allow request below the limit", function()
```
Comment (catbro666) [inline spec/03-plugins/38-basic-rate-limiting/02-access_spec.lua:58]: We can add one more test to verify that the counter get reset after the ttl.

Comment (mrwanny) [reply spec/03-plugins/38-basic-rate-limiting/02-access_spec.lua:58]: very good idea

File: kong/plugins/basic-rate-limiting/handler.lua (hunk 1 above)
Comment (catbro666) [inline kong/plugins/basic-rate-limiting/handler.lua:14]: style: indent 2 spaces

**Review by catbro666 — APPROVED** [review]
Overall LGTM. Do we need to add an entry in `CHANGELOG.md` or is it for internal use only?

**Comment by hbagdi** [comment]
Please hold.

**Comment by hbagdi** [comment]
No plans to ship this just yet.

**Pull request closed without merging**